### ISSUE TITLE
[DTOs] [Runtime] Communicate failure

### ISSUE THREAD

**Issue by rinswind** [pkriens/cdi#30]
**Issue**
The `CdiContainer` and/or the DTOs must communicate failures
- Failures during initialization: e.g. definition errors.
- Failures at runtime: e.g. exceptions during component context creation.
    - As these can happen more than once and accumulate store just the last one?
    - ... or maybe fail the container entirely on first failure?
- Any other failures?

**Suggestion** 
- Add `String[] faiures;` or similar to `ContainerDTO` and/or `ContainerActivationDTO`
- Add `Exception getFailure()` to `CdiContainer` for the stack trace.
- Require the logging of failures to `LogService` if available.

**Comment by rotty3000** [comment]
Yup, we haven't added any failure DTOs yet! We've only represented the healthy state. But we needed to know the healthy state design before having the error state. I think we are close now.

**Comment by pkriens** [comment]
Since the unhealthy state is infinite (and often in impls details) I think the log service is the place to go. We already indicate the state of the container and that has failure. We could add expand the ContainerState for common problems and add a string for details? However, in the end the details MUST be in the log.

**Comment by rotty3000** [comment]
perhaps, but in order to have a little bit of parity with other specs, we need Failure DTOs because late inspection of the state of the container can be tedious to dig out of a log.

**Comment by pkriens** [comment]
Tend to disagree because there are so many possible ways thing can fail. I think a summary should suffice maybe combined with log statements that contain the Container id?

**Comment by rinswind** [comment]
We have major issues with catching failure in DS.
It is very useful to have a persistent place to extract the failure preferably with a console command, when digging into a failed OSGi runtime.

**Comment by pkriens** [comment]
in that case it would be nice to see what kind of failures we're talking about. In my experiences this is often more how we like the world to be than how the world actually is. Failures are messy and rarely fit nice DTOs or they could already be handled.

**Comment by rinswind** [comment]
Notes form the EEG f2f

# Rules
- Generally when each level builds the next level there is opportunity for failure
```
nothing --> 
    container -->
        lifecycle -->
            component -->
                activation`
````
- Publish errors caused by the user
    - Not bugs in the extender
- Publish persistent errors
    - Remain until external change causes a level to be rebuilt
    - Not error streams like failures to create a service object
- All other errors go to the Log Service

# Error format
- Leave room for multiple errors
    - E.g. all definition errors
- Store the exceptions as strings

# Result
- Candidate error detection sites
```
nothing --(bad bundle errors)--> 
    container -->
        lifecycle -->
            component --(activation errors)-->
                activation`
````
- Places to store them
    - bad bundle errors: in a failed `container` object
    - bad dependency errors: in failed `activation` object
    - I.e. make the next level even if failed and store the errors there.

# Activation errors are under question
What can they be?
- The extender can detect failures to coerse the configuration with the converter.
- The extender can detect failures to consume services.
- It is probably not a good idea to store failures of creating the CDI contexts, because prototype/bundle create many (fail the activation on first failure?).

**Comment by rinswind** [comment]
Removing `action-required` because we have not decided on this
